### PR TITLE
refactor(registry): Small optimization of Node Provider Rewards

### PR DESCRIPTION
**Pull request by max-dfinity**

### CODE REVIEW

File: rs/registry/canister/src/get_node_providers_monthly_xdr_rewards.rs [hunk 1]
```diff
@@ -35,16 +38,8 @@ impl Registry {
...
-            if key.starts_with(NODE_OPERATOR_RECORD_KEY_PREFIX.as_bytes()) {
-                let value = values.back().unwrap();
-                if value.deletion_marker {
-                    continue;
-                }
-                let node_operator = NodeOperatorRecord::decode(value.value.as_slice()).unwrap();
-                let node_operator_id = PrincipalId::try_from(
-                    &node_operator.node_operator_principal_id,
-                )
+        for (key, node_operator) in get_key_family_iter(&self, NODE_OPERATOR_RECORD_KEY_PREFIX) {
```
Comment (max-dfinity) [inline rs/registry/canister/src/get_node_providers_monthly_xdr_rewards.rs:41]: Note for reviewer: These are the relevant changes, everything else is whitespace

File: rs/registry/canister/src/get_node_providers_monthly_xdr_rewards.rs [hunk 2]
```diff
@@ -64,124 +57,123 @@ impl Registry {
...
-                    println!(
-                        "NodeProvider {} reward for all {} {} nodes in {} DC: reward {}",
-                        node_provider_id,
-                        node_count,
-                        node_type.as_str(),
-                        node_operator.dc_id,
-                        dc_reward,
-                    );
-                    *np_rewards += dc_reward;
-                }
```
Comment (max-dfinity) [inline rs/registry/canister/src/get_node_providers_monthly_xdr_rewards.rs:184]: this closing bracket is other relevant change

File: rs/registry/canister/src/mutations/node_management/common.rs [hunk 3]
```diff
@@ -226,6 +226,13 @@ fn get_key_family<T: prost::Message + Default>(
     registry: &Registry,
     prefix: &str,
 ) -> Vec<(String, T)> {
+    get_key_family_iter(registry, prefix).collect()
+}
+
+pub(crate) fn get_key_family_iter<T: prost::Message + Default>(
```
Comment (max-dfinity) [inline rs/registry/canister/src/mutations/node_management/common.rs:232]: new method to more efficiently iterate over all entries of a particular type than what we were doing, but without the memory overhead of collecting them into a vector all at once.

File: rs/registry/canister/src/mutations/node_management/common.rs [hunk 4]
```diff
@@ -226,6 +226,13 @@ fn get_key_family<T: prost::Message + Default>(
     registry: &Registry,
     prefix: &str,
 ) -> Vec<(String, T)> {
+    get_key_family_iter(registry, prefix).collect()
+}
+
+pub(crate) fn get_key_family_iter<'a, T: prost::Message + Default>(
```
Comment (anchpop) [inline rs/registry/canister/src/mutations/node_management/common.rs:232]: NCR: why not put this in

```
impl<'a> Registry {
    pub fn key_family<'a, T: prost::Message + Default>(&'a self, &'a str) -> impl Iterator<Item = (String, T)> + 'a { 
        ...
    }
}
```

Also, I don't know if we need to write `get_` or `_iter`, the rust standard library doesn't do this and I think it's clear enough what it's doing without it (e.g. to get an iterator over the values of a hashmap you just do `mymap.values()`).

**Review by anchpop — APPROVED** [review]

File: rs/registry/canister/src/mutations/node_management/common.rs (hunk 4 above)
Comment (max-dfinity) [reply rs/registry/canister/src/mutations/node_management/common.rs:232]: There's already a "get_key_family" which just returns all the key/value pairs, and I wanted to differentiate between that and a function that returns an iterator.

The main change here is to stop iterating over all the registry keys like we currently do, which is quite expensive.

**Pull request merged**